Add parameter to select pages to be exported by CLI

PgBiel opened the pull request:
Closes https://github.com/typst/typst/issues/3095

Adds the `--pages` parameter to `typst compile` and `typst watch` which allows selecting which pages should be exported by the CLI. Supports all export formats (PDF, PNG, SVG).

- **Syntax:**
   - **Simple:** `typst compile --pages 5,6,11 a.typ a.pdf` to export a PDF with just pages 5, 6, 11.
   - **Ranges:** `typst watch --pages 11-13,15- a.typ a-{n}.png` to export pages 11 to 13 (inclusive), as well as any pages after page 15. **(Not yet implemented)**

- **Implementation:**
   - Had to refactor `typst-pdf` in several places to consider non-exported pages. Changed `ctx.pages` to `Vec<Option<EncodedPage>>` to preserve the relation between indices and real page numbers.
   - For PNG and SVG, the implementation was trivial, since each page is a separate exported file.

- **Decisions:**
  - [ ] (PDF) Should non-exported headings be excluded from PDF bookmarks? Currently, I have excluded them, as this seems to make sense (otherwise they would be broken links).
  - [ ] (PDF) The PDF's `xmp.num_pages` metadata should only contain the amount of exported pages (and should thus always correspond to the number of pages in the actual PDF, not necessarily in the original Typst document), correct? (I have assumed so for now)
  - [ ] (PNG/SVG) When using `exported-name-{n}.png` (for example), should `{n}` correspond to the real page number in the original typst document? (i.e. with `--pages 5,6,7` you'd get `name-5.png`, `name-6.png` and `name-7.png` instead of `name-1.png`, `name-2.png` and `name-3.png`) I've assumed so for now as well, as I'd say this would be expected.

- **TODO:**
   - [ ] Support ranges in `--pages`
   - [ ] Recheck PDF changes
   - [ ] Undraft

PgBiel commented on crates/typst-pdf/src/lib.rs:
```diff
@@ -55,13 +76,17 @@ use crate::pattern::PdfPattern;
     document: &Document,
     ident: Smart<&str>,
     timestamp: Option<Datetime>,
+    page_ranges: Option<Vec<Range<NonZeroUsize>>>,
```
Yeah, we can't use it in `args.rs` sadly, but with the refactor the `pdf()` call looks simpler now.

laurmaedje commented on crates/typst-cli/src/compile.rs:
```diff
@@ -171,7 +173,17 @@ fn export_pdf(document: &Document, command: &CompileCommand) -> StrResult<()> {
     );
-    let buffer = typst_pdf::pdf(document, Smart::Auto, timestamp);
+    let exported_page_ranges: Option<PageRanges> =
+        command.pages.as_ref().map(|export_ranges| {
+            PageRanges::new(
+                export_ranges
+                    .iter()
+                    .cloned()
+                    .map(PageRangeArgument::into_range)
+                    .collect(),
+            )
+        });
```
We could deduplicate this into a `CompileCommand::pages` in this file.

laurmaedje commented on crates/typst-pdf/src/outline.rs:
```diff
@@ -18,7 +18,17 @@ pub(crate) fn write_outline(ctx: &mut PdfContext) -> Option<Ref> {
+
     for elem in elements.iter() {
+        if let Some(page_ranges) = &ctx.exported_pages {
+            if !page_ranges.page_included(
+                ctx.document.introspector.page(elem.location().unwrap()).get() - 1,
+            ) {
```
Maybe we could have `includes_page` and `includes_page_index`, so that we don't need to have `-1` sprinkled randomly.

PgBiel commented on crates/typst-cli/src/args.rs:
```diff
@@ -271,6 +290,54 @@ fn parse_input_pair(raw: &str) -> Result<(String, String), String> {
+    fn from_str(value: &str) -> Result<Self, Self::Err> {
+        match value
+            .split('-')
+            .map(|part| part.trim_matches(' '))
```
At first I wasn't doing any trimming, until I noticed the web app does allow spaces, and I wanted to keep it consistent with the web app as requested. However, I wanted to be cautious regarding making the API overly flexible, so I kept it at just spaces. Though I think there isn't much of a problem, so I have switched it to `trim()`.

PgBiel commented:
I noticed something interesting: without page labels (custom page numbering), the PDF page numbers (in the outline or otherwise) do not correspond to their real page numbers in the original Typst document. Consider the document below compiled with `--pages "1,2,3,4,8-10"`, where each page has a "Bruh" heading:

1. Without page labels - Okular PDF reader shows 5,6,7 (real PDF page number) instead of 8,9,10 (page number in Typst).
     ![image](https://github.com/typst/typst/assets/9021226/f3bbf16b-d35f-4292-a1f6-cc3b3bab1213)

2. With page labels - Okular PDF reader shows `h.`, `i.` and `j.`, matching how Typst displays page numbers, although the real PDF page numbers are kept and displayed separately. 
     ![image](https://github.com/typst/typst/assets/9021226/32afaf1c-9632-4159-9d5c-2f77e9322287)


Same happens for other PDF readers, e.g. in Firefox. The default, if Typst doesn't specify custom page numbering labels, is to just display the real PDF page number instead of the Typst page number.

I wonder if there's much that can be done about this, e.g. always specify page labels if some pages are excluded from exporting. But perhaps we could consider this to be "intended behavior" in a way. Or maybe it's just not worth it to fix this / could fix later (doesn't seem to be that bad of a problem).

laurmaedje commented:
> always specify page labels if some pages are excluded from exporting

That might make sense indeed, especially since it already happens if a numbering is specified.

PgBiel commented:
Alright, I've implemented this, though there are two important notes (and thus potential design decisions):

1. I noticed that, when page numbering was enabled by the Typst document, the page numbers used for the labels are the logical page numbers, not the real page numbers. When page numbering is disabled, however, the page numbers used are always the real page numbers (since just the PDF's final page numbers are used, given no labels are added). In my implementation of adding labels to pages without numbering after non-exported pages, I added labels with the real page numbers, not logical, for consistency with the case of no page numbering. But let me know if that wasn't the right choice.

2. Consider the situation where the first N pages were exported, but page N + 1 wasn't exported, so pages N + 2 and onwards have PDF page numbers which don't correspond to their real page numbers in the typst document (as shown in my last reply). The solution I implemented was to add labels to pages N + 2  onwards with their respective real page numbers in Typst. However, there was one thing I wasn't sure about: should I have added labels to the first N pages too (redundant labels, equal to the final PDF page numbers in every way) just for consistency? Initially I experimented with doing so in https://github.com/typst/typst/pull/4039/commits/0ffac2e0aa1719bb37655138e6f24331d4e4d115, but ended up reverting it because it felt unnecessary. Therefore, only pages N + 2 onwards are labeled (just to fix the problem of "drifting" page numbers after non-exported pages). Let me know if this wasn't the right choice as well.

laurmaedje commented:
Your choices sound fine. Thank you!
